lib/cretonne/src/ssa.rs

```diff
@@ -0,0 +1,520 @@
+            // The block has multiple predecessors, we register the ebb argument as the current
+            // definition for the variable.
+            Some((val, sealed, ebb)) => {
+                self.def_var(var, val, block);
+                if sealed {
+                    self.seal_ebb_header_block(ebb, dfg)
+                }
+                val
+            }
+        }
+
+    }
+
+    /// Declares a new basic block belonging to the body of a certain `Ebb` and having `pred`
+    /// as a predecessor. `pred` is the only predecessor of the block and the block is sealed
+    /// at creation.
+    ///
+    /// To declare a `Ebb` header block, see `declare_ebb_header_block`.
+    pub fn declare_ebb_body_block(&mut self, ebb: Ebb, pred: Block) -> Block {
+        self.blocks.push(BlockData::EbbBody(pred, ebb))
+    }
+
+    /// Declares a new basic block at the beginning of an `Ebb`. No predecessors are declared
+    /// here and the block is not sealed.
+    /// Predecessors have to be added with `declare_ebb_predecessor`.
+    pub fn declare_ebb_header_block(&mut self, ebb: Ebb) -> Block {
+        let block = self.blocks
+            .push(BlockData::EbbHeader(HashMap::new(), false, ebb, Vec::new()));
+        self.ebb_headers.insert((ebb, block));
+        block
+    }
+
+    /// Declares a new predecessor for an `Ebb` header block and record the branch instruction
+    /// of the predecessor that leads to it.
+    ///
+    /// Note that the predecessor is a `Block` and not an `Ebb`. This `Block` must be filled
+    /// before added as predecessor. Note that you must provide no jump arguments to the branch
+    /// instruction when you create it since `SSABuilder` will fill them for you.
+    pub fn declare_ebb_predecessor(&mut self, ebb: Ebb, pred: Block, inst: Inst) {
+        let header_block = match self.ebb_headers.get(ebb) {
+            None => panic!("you are declaring a predecessor for an ebb not declared"),
+            Some(&(_, header)) => header,
+
+        };
+        self.blocks[header_block].add_predecessor(pred, inst)
+    }
+
+    /// Completes the global value numbering for an `Ebb`, all of its predecessors having been
+    /// already sealed.
+    ///
+    /// This method modifies the function's `Layout` by adding arguments to the `Ebb`s to
+    /// take into account the Phi function placed by the SSA algorithm.
+    pub fn seal_ebb_header_block(&mut self, ebb: Ebb, dfg: &mut DataFlowGraph) {
+        let block = match self.ebb_headers.get(ebb) {
+            None => panic!("this ebb has no block header defined"),
+            Some(&(_, block)) => block,
+        };
+        // TODO: find a way to not allocate vectors
+        let (predecessors, undef_vars): (Vec<(Block, Inst)>, Vec<(Variable, Value)>) =
+            match self.blocks[block] {
+                BlockData::EbbBody(_, _) => panic!("this should not happen"),
+                BlockData::EbbHeader(ref predecessors, _, _, ref undef_vars) => {
```

Comment:
Would it make sense to have a separate function be the public API, which can `debug_assert!(!sealed)` or so, with the private one doing the main work? I think you could also move the `self.ebb_headers.get(ebb)` lookup into the public one, because the callsite in `use_var` already knows the `Block`.